src/qibolab/platforms/multiqubit.py

```diff
@@ -45,17 +140,45 @@ def execute_pulse_sequence(self, sequence: PulseSequence, nshots=None):
             if "readout" in roles[name]:
                 if not instrument_pulses[name].is_empty:
                     if not instrument_pulses[name].ro_pulses.is_empty:
+                        if all([pulse.serial in changed for pulse in instrument_pulses[name].ro_pulses]):
+                            # FIXME: for precision sweep in resonator spectroscopy
+                            # change necessary to perform precision sweep
+                            # TODO: move this to instruments (ask Alvaro)
+                            # TODO: check if this will work with multiplex
+                            for sequencers in self.instruments[name]._sequencers.values():
+                                for sequencer in sequencers:
+                                    sequencer.pulses = instrument_pulses[name].ro_pulses
                         results = self.instruments[name].acquire()
                         existing_keys = set(acquisition_results.keys()) & set(results.keys())
                         for key, value in results.items():
                             if key in existing_keys:
                                 acquisition_results[key].update(value)
                             else:
                                 acquisition_results[key] = value
-        return acquisition_results
+
+        # change back the frequency of the pulses
+        for name in self.instruments:
+            roles[name] = self.settings["instruments"][name]["roles"]
+            if "readout" in roles[name]:
+                instrument_pulses[name] = sequence.get_channel_pulses(*self.instruments[name].channels)
+                for pulse in instrument_pulses[name]:
+                    if pulse.serial in changed:
+                        pippo = acquisition_results[pulse.serial]
+                        # if abs(pulse.frequency) > 300e6:
+                        pulse.frequency += self.get_lo_readout_frequency(pulse.qubit)
+                        acquisition_results[pulse.serial] = pippo
+            if "control" in roles[name]:
+                instrument_pulses[name] = sequence.get_channel_pulses(*self.instruments[name].channels)
+                for pulse in instrument_pulses[name]:
+                    if pulse.serial in changed:
+                        pulse.frequency += self.get_lo_drive_frequency(pulse.qubit)
+
+        for ro_pulse in ro_pulses.values():
+            data[ro_pulse.serial] = ExecutionResults.from_components(*acquisition_results[key])
+            data[ro_pulse.qubit] = ExecutionResults.from_components(*acquisition_results[key])
```

Comment:
I think here we should have both dictionaries reference to the same object 
```suggestion
            data[ro_pulse.qubit] = data[ro_pulse.serial] = ExecutionResults.from_components(*acquisition_results[key])
```

Reply:
If you don't need to copy it, that is for sure true.
And in case you need to copy, better to do it explicitly, deciding also if the copy should be shallow, `{}.copy()`, or deep, `copy.deepcopy({})`. But do not recompute it twice, in case you assign to a temporary variable :)

Reply:
(re-retrieving the dict entry is effectively the same: you consume less memory at the price of taking the element back again, that is $O(1)$ for dicts)

Reply:
Thanks @AleCandido and @stavros11. I will try not make this mistake again :)